Git: 2.48.1 -> 2.49.0

me-and opened the pull request:
A few changes in addition to the straightforward version bump:

- Update the src url handling to cope with release candidate version strings. This makes no difference for the actual releases, but made testing the release candidates as they came out easier.
- Fix git-send-email-honor-PATH.patch to work after the upstream change in <https://github.com/gitster/git/commit/1f010d6bdf756129db13d1367c888aa4153f6d87>.
- Use make flags in the `postBuild` and `postInstall` hooks in a way that's consistent with how the flags are handled in the main stdenv build phase. This is necessary because the inconsistent handling of `PERL_PATH` flags was causing build failures in this release.
- Swap out zlib for the newly supported and faster zlib-ng. This adds about 176kiB to the closure size. 

Release notes: https://github.com/git/git/blob/v2.49.0/Documentation/RelNotes/2.49.0.adoc

## Things done

- Built:
  - x86_64-linux
    - [x] `git`
    - [x] `git.passthru.tests`
    - [x] `gitMinimal`
    - [x] `gitMinimal.passthru.tests`
    - [x] `gitFull`
    - [x] `gitFull.passthru.tests`
    - [x] `gitSVN`
    - [x] `gitSVN.passthru.tests`
    - [x] `nixosTests.gitdaemon`
    - [x] `tests.fetchgit`
  - aarch64-linux
    - [x] `git`
    - [ ] `git.passthru.tests`
    - [x] `gitMinimal`
    - [ ] `gitMinimal.passthru.tests`
    - [ ] `gitFull`
    - [ ] `gitFull.passthru.tests`
    - [x] `gitSVN`
    - [ ] `gitSVN.passthru.tests`
    - [ ] `nixosTests.gitdaemon`
    - [x] `tests.fetchgit`
  - [ ] x86_64-darwin (I don't have access to a test system)
  - [ ] aarch64-darwin (I don't have access to a test system)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] ~(Package updates) Added a release notes entry if the change is major or breaking~ _no major or breaking changes_
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

me-and committed:
git: use make flags for all make calls

Rather than have the default stdenv builder call `make` with one set of
flags, then call `make` with a variety of other flags in the postBuild
and postInstall phases, build the same list of flags as the stdenv build
script, then use that flag list as an argument to calls to make.

This means the flags are more consistent across all calls.  In
particular, that means the documentation build doesn't fail because it
doesn't have PERL_PATH set when building Git v2.49.0-rc0.  It also means
more build steps are run using as many cores as are available for the
main build stages, and the install stages control their parallelism with
consistent use of the `enableParallelInstalling` flag (which is now also
enabled).

me-and committed:
git: 2.48.1 -> 2.49.0

As well as the straightforward version bump, update the src url to
handle release candidate version strings, and fix
git-send-email-honor-PATH.patch to work after the upstream change in
1f010d6bdf (doc: use .adoc extension for AsciiDoc files, 2025-01-20).

leiserfg commented:
Should we replace zlib with zlib-ng here?
https://about.gitlab.com/blog/2025/03/14/whats-new-in-git-2-49-0/#introduction-of-zlib-ng

me-and committed:
git: use zlib-ng rather than zlib

Git v2.49.0 added support for using the faster zlib-ng library, and
benchmarks by GitLab show it's notably faster, so use it for our builds
too.

The different zlib implementation does change the contents of the .git
directory, which means the hashes for fetchgit tests that leave the .git
directory in place are expected to change.

https://about.gitlab.com/blog/2025/03/14/whats-new-in-git-2-49-0/#introduction-of-zlib-ng

me-and commented:
> Should we replace zlib with zlib-ng here? https://about.gitlab.com/blog/2025/03/14/whats-new-in-git-2-49-0/#introduction-of-zlib-ng

Good shout. I've just done some quick tests and it seems to work as a drop-in replacement. The closure, even for gitMinimal, still includes the regular zlib library, because it's pulled in by curl, but it's only an extra 176 kiB on the total closure size.

I'll update the PR now with a new commit to add zlib-ng; I'm just going to fix it unless someone thinks there's a compelling argument for making this overridable.

me-and committed:
tests.fetchgit: test deep clone without .git

The deepClone argument to fetchgit implies leaveDotGit.  Make the
"submodule-deep" fetchgit test more useful by having it remove the .git
directory, as recommended in the fetchgit documentation, so it is
distinguished from the "submodule-leave-git-deep" test.

philiptaron reviewed (approved):
LGTM. Doing a build on x86_64-linux. Would appreciate build and test reports from @NixOS/darwin-core, if your machines aren't doing anything else across the weekend. Or a delegation to some other set of folks with Mac resources.

me-and commented:
(I'll keep updating the build results table in the OP as things complete, but my Raspberry Pi is struggling a bit so aarch64 builds are likely to take a while...)

me-and commented:
> Woah cool speedup - shouldn't we replace zlib everywhere?

Switching libraries changed the hashes in the `fetchgit` tests that kept the `.git` directory, so the zlib-ng outputs aren't byte-for-byte identical to the zlib ones. I expect that's fine in the vast majority of cases &ndash; changing the dependencies would change the derivation output anyway &ndash; but I wouldn't want to assume those changes don't cause more significant issues for at least some packages, so it's probably safer to approach this on a package-by-package basis.

That said, a project to switch over on a package-by-package basis seems like it could produce a decent performance increase across Nixpkgs as a whole!

philiptaron commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 385402 --package git.passthru.tests --package gitSVN.passthru.tests --package git --package gitFull.passthru.tests --package gitMinimal --package nixosTests.gitdaemon --package gitMinimal.passthru.tests`

---
### `x86_64-linux`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>git.passthru.tests.buildbot-integration</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 test built:</summary>
  <ul>
    <li>nixosTests.gitdaemon</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 20 packages built:</summary>
  <ul>
    <li>git</li>
    <li>git.debug (git.debug.debug, git.debug.doc)</li>
    <li>git.doc (git.doc.debug, git.doc.doc)</li>
    <li>git.passthru.tests.dumb-http-signed-tag</li>
    <li>git.passthru.tests.leave-git</li>
    <li>git.passthru.tests.simple</li>
    <li>git.passthru.tests.sparseCheckout</li>
    <li>git.passthru.tests.sparseCheckoutNonConeMode</li>
    <li>git.passthru.tests.submodule-deep</li>
    <li>git.passthru.tests.submodule-leave-git</li>
    <li>git.passthru.tests.submodule-leave-git-deep</li>
    <li>git.passthru.tests.submodule-simple</li>
    <li>gitFull.passthru.tests.withInstallCheck</li>
    <li>gitFull.passthru.tests.withInstallCheck.debug (gitFull.passthru.tests.withInstallCheck.debug.debug, gitFull.passthru.tests.withInstallCheck.debug.doc)</li>
    <li>gitFull.passthru.tests.withInstallCheck.doc (gitFull.passthru.tests.withInstallCheck.doc.debug, gitFull.passthru.tests.withInstallCheck.doc.doc)</li>
    <li>gitMinimal</li>
    <li>gitMinimal.debug (gitMinimal.debug.debug)</li>
    <li>gitSVN.passthru.tests.withInstallCheck</li>
    <li>gitSVN.passthru.tests.withInstallCheck.debug (gitSVN.passthru.tests.withInstallCheck.debug.debug, gitSVN.passthru.tests.withInstallCheck.debug.doc)</li>
    <li>gitSVN.passthru.tests.withInstallCheck.doc (gitSVN.passthru.tests.withInstallCheck.doc.debug, gitSVN.passthru.tests.withInstallCheck.doc.doc)</li>
  </ul>
</details>

emilazy commented:
> Switching libraries changed the hashes in the `fetchgit` tests that kept the `.git` directory, so the zlib-ng outputs aren't byte-for-byte identical to the zlib ones. I expect that's fine in the vast majority of cases – changing the dependencies would change the derivation output anyway – but I wouldn't want to assume those changes don't cause more significant issues for at least some packages, so it's probably safer to approach this on a package-by-package basis.

Breaking `leaveDotGit` FOD hashes seems scary. I know they’re not super reliable to begin with, but I worry about this breaking things silently throughout the tree (outdated FOD hashes aren’t really caught by Hydra) and in external projects. We might want to have `fetchgit` use Git with `zlib` instead when `leaveDotGit` is on?

me-and commented:
> > Switching libraries changed the hashes in the `fetchgit` tests that kept the `.git` directory, so the zlib-ng outputs aren't byte-for-byte identical to the zlib ones. I expect that's fine in the vast majority of cases – changing the dependencies would change the derivation output anyway – but I wouldn't want to assume those changes don't cause more significant issues for at least some packages, so it's probably safer to approach this on a package-by-package basis.
> 
> Breaking `leaveDotGit` FOD hashes seems scary. I know they’re not super reliable to begin with, but I worry about this breaking things silently throughout the tree (outdated FOD hashes aren’t really caught by Hydra) and in external projects. We might want to have `fetchgit` use Git with `zlib` instead when `leaveDotGit` is on?

We've already broken `leaveDotGit` FOD hashes recently, in #372784. Based both on the discussion there, and the fact that (as best I'm aware) nothing has exploded horrendously in the past month, I'm not worried about this change making anything worse than it already is.

me-and commented:
I'm currently building `git.passthru.tests.buildbot-integration` to check if the failure @philiptaron saw is pre-existing on staging, but I've re-confirmed that test is passing on the tip of my PR branch.

emilazy commented:
> We've already broken `leaveDotGit` FOD hashes recently, in #372784. Based both on the discussion there, and the fact that (as best I'm aware) nothing has exploded horrendously in the past month, I'm not worried about this change making anything worse than it already is.

Fair enough! But FWIW, the problem is generally that breaking FOD hashes *doesn’t* blow up: since the cache already has the old version, Hydra never tries to rebuild it, and it’s only years down the line when someone tried to reproduce a build from scratch that it breaks. (But I agree that in this case it seems like it’s probably fine, since `leaveDotGit` is so cursed already.)

leiserfg commented:
Isn't leaveDotGit considered an impurity?

me-and commented:
> Isn't leaveDotGit considered an impurity?

The point of fixed-output derivations is that they're the product of things other than their formal inputs. I don't think using `leaveDotGit` is any more or less impure than relying on, say, downloading a repository from the network in general.

It's clearly _fragile_ – as we see here, using `leaveDotGit` makes it much more likely that the hash of the derivation will be changed in circumstances where that's (presumably) not desirable – but that doesn't make it impure.

leiserfg commented:
I'm not sure,  if the src of a derivation points to a commit and uses the `.git` for extracting the revision description, that can change without changing the commit hash as one can create a tag pointing to a commit and that does not changes the commit, nor the hash isn't it?

philiptaron commented:
I'd like to merge this in the next couple days.

me-and commented:
@philiptaron I've just rebuilt `git.passthru.tests.buildbot-integration` with my commits rebased to the tip of the staging branch, and that builds for me. I don't know if you hit an intermittent build failure or just got unlucky with that test being temporarily broken on the tip of staging.

philiptaron commented:
Yeah, I'm not worried. I would like to have had a darwin build, but I suspect it'll all come up roses anyway.

philiptaron commented:
https://github.com/NixOS/nixpkgs/runs/38862526394 indeed it did come up roses.
